repository-s3/src/main/java/io/aiven/elasticsearch/repositories/s3/S3OutputStream.java

```diff
@@ -0,0 +1,165 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.elasticsearch.repositories.s3;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.model.AbortMultipartUploadRequest;
+import com.amazonaws.services.s3.model.CompleteMultipartUploadRequest;
+import com.amazonaws.services.s3.model.InitiateMultipartUploadRequest;
+import com.amazonaws.services.s3.model.ObjectMetadata;
+import com.amazonaws.services.s3.model.PartETag;
+import com.amazonaws.services.s3.model.UploadPartRequest;
+import org.elasticsearch.common.Strings;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class S3OutputStream extends OutputStream {
+
+    private final Logger logger = LoggerFactory.getLogger(S3OutputStream.class);
+
+    private final AmazonS3 client;
+
+    private final ByteBuffer byteBuffer;
+
+    private final String bucketName;
+
+    private final String key;
+
+    private MultipartUpload multipartUpload;
+
+    private final int partSize;
+
+    public S3OutputStream(final String bucketName,
+                           final String key,
+                           final int partSize,
+                           final AmazonS3 client) {
+        this.bucketName = bucketName;
+        this.key = key;
+        this.client = client;
+        this.partSize = partSize;
+        this.byteBuffer = ByteBuffer.allocate(partSize);
+    }
+
+    @Override
+    public void write(final int b) throws IOException {
+        write(new byte[]{(byte) b}, 0, 1);
+    }
+
+    @Override
+    public void write(final byte[] b, final int off, final int len) throws IOException {
+        if (Objects.isNull(b) || b.length == 0) {
+            return;
+        }
+        if (Objects.isNull(multipartUpload)) {
+            multipartUpload = newMultipartUpload();
+        }
+        final var source = ByteBuffer.wrap(b, off, len);
+        while (source.hasRemaining()) {
+            final var transferred = Math.min(byteBuffer.remaining(), source.remaining());
+            final var offset = source.arrayOffset() + source.position();
+            byteBuffer.put(source.array(), offset, transferred);
+            source.position(source.position() + transferred);
+            if (!byteBuffer.hasRemaining()) {
+                flushBuffer(0, partSize, partSize);
+            }
+        }
+    }
+
+    private MultipartUpload newMultipartUpload() throws IOException {
+        logger.debug("Create new multipart upload request");
+        final var initialRequest = new InitiateMultipartUploadRequest(bucketName, key);
+        final var omd = new ObjectMetadata();
+        omd.setSSEAlgorithm(ObjectMetadata.AES_256_SERVER_SIDE_ENCRYPTION);
```

Comment:
We don't use server-side encryption, or do we?

Reply:
We have it by default. The ES plugin supports it explicitly (you need to specify so you need server-side encryption or not)

Reply:
SSE is encryption done on the AWS side with AWS-managed keys. I'm a bit puzzled to see why do we need it and how we use it, as we do encryption on the client side. Could you plz elaborate?

Reply:
true ... will remove this code. We do not need it.